ComponentKit/TransactionalDataSources/Common/Internal/CKTransactionalComponentDataSourceChangesetVerification.mm

```diff
@@ -0,0 +1,161 @@
+   As long as each item is located within the bounds of its section the changeset is valid.
+   */
+  NSMutableDictionary<NSNumber *, NSMutableIndexSet *> *itemsToRemove = [NSMutableDictionary new];
+  [changeset.removedItems enumerateObjectsUsingBlock:^(NSIndexPath *_Nonnull indexPath, BOOL * _Nonnull stop) {
+    const NSInteger section = indexPath.section;
+    const NSInteger item = indexPath.item;
+    if (section >= sectionCounts.count
+        || section < 0
+        || item >= [sectionCounts[section] integerValue]) {
+      invalidChangeFound = YES;
+      *stop = YES;
+    } else {
+      if (!itemsToRemove[@(section)]) {
+        itemsToRemove[@(section)] = [NSMutableIndexSet indexSet];
+      }
+      [itemsToRemove[@(section)] addIndex:item];
+    }
+  }];
+  if (invalidChangeFound) {
+    return CKBadChangesetOperationTypeRemoveRow;
+  } else {
+    [itemsToRemove enumerateKeysAndObjectsUsingBlock:^(NSNumber * _Nonnull section, NSMutableIndexSet * _Nonnull indexSet, BOOL * _Nonnull stop) {
+      sectionCounts[[section integerValue]] = @([sectionCounts[[section integerValue]] integerValue] - [indexSet count]);
+    }];
+  }
+  /*
+   Removed sections
+   Section counts may not immediately reflect removals as order is not guaranteed and may result in a false positive.
+   As long as each section is located within the bounds of all sections the changeset is valid.
+   */
+  NSMutableIndexSet *sectionsToRemove = [NSMutableIndexSet indexSet];
+  [changeset.removedSections enumerateIndexesUsingBlock:^(NSUInteger section, BOOL * _Nonnull stop) {
+    if (section >= sectionCounts.count) {
+      invalidChangeFound = YES;
+      *stop = YES;
+    } else {
+      [sectionsToRemove addIndex:section];
+    }
+  }];
+  if (invalidChangeFound) {
+    return CKBadChangesetOperationTypeRemoveSection;
+  } else {
+    [sectionCounts removeObjectsAtIndexes:sectionsToRemove];
+  }
+  /*
+   Inserted sections
+   Section counts may immediately reflect insertions as they are guaranteed to be contiguous by virtue of NSIndexSet.
+   As long as each section is located within the bounds of all sections the changeset is valid.
+   */
+  [changeset.insertedSections enumerateIndexesUsingBlock:^(NSUInteger section, BOOL * _Nonnull stop) {
+    if (section > sectionCounts.count) {
+      invalidChangeFound = YES;
+      *stop = YES;
+    } else {
+      [sectionCounts insertObject:@0 atIndex:section];
+    }
+  }];
+  if (invalidChangeFound) {
+    return CKBadChangesetOperationTypeInsertSection;
+  }
+  /*
+   Inserted items
+   Section counts may immediately reflect insertions as they are guaranteed to be contiguous by virtue of sorting the index paths.
+   As long as each item is located within the bounds of its section the changeset is valid.
+   */
+  [sortedIndexPaths(changeset.insertedItems.allKeys) enumerateObjectsUsingBlock:^(NSIndexPath * _Nonnull indexPath, NSUInteger index, BOOL * _Nonnull stop) {
```

Comment:
This case is a bit weird since there is no guaranteed order of item insertions. I chose to sort the index paths so that they may be inserted contiguously. If for some reason there's a "gap", or an item is inserted outside the bounds of the current section, the insertion will be invalid.